app/ide-desktop/lib/dashboard/src/pages/dashboard/Dashboard.tsx

```diff
@@ -320,40 +361,50 @@ export default function Dashboard(props: DashboardProps) {
     }
   }, [inputBindings])
 
-  const doOpenEditor = React.useCallback(() => {
-    setPage(TabType.editor)
-  }, [setPage])
-
-  const doCloseEditor = React.useCallback(
-    (id: backendModule.ProjectId) => {
-      if (id === projectStartupInfo?.projectAsset.id) {
-        setProjectStartupInfo(currentInfo => {
-          if (id === currentInfo?.projectAsset.id) {
-            setPage(TabType.drive)
-            return null
-          } else {
-            return currentInfo
-          }
-        })
-      }
-    },
-    [projectStartupInfo?.projectAsset.id, setPage]
-  )
-
-  const doRemoveSelf = React.useCallback(() => {
-    if (projectStartupInfo?.projectAsset != null) {
-      const id = projectStartupInfo.projectAsset.id
-      dispatchAssetListEvent({ type: AssetListEventType.removeSelf, id })
-      setProjectStartupInfo(null)
+  const doCloseAllProject = eventCallbacks.useEventCallback(() => {
+    for (const launchedProject of launchedProjects) {
+      doCloseProject(launchedProject)
     }
-  }, [projectStartupInfo?.projectAsset, dispatchAssetListEvent])
+  })
 
-  const onSignOut = React.useCallback(() => {
-    if (page === TabType.editor) {
-      setPage(TabType.drive)
+  const doOpenProject = eventCallbacks.useEventCallback((project: Project) => {
+    // since we don't support multitabs, we need to close opened project first
+    if (launchedProjects.length > 0) {
+      doCloseAllProject()
     }
-    setProjectStartupInfo(null)
-  }, [page, setPage])
+
+    openProjectMutation.mutate(project)
+    addLaunchedProject(project)
+  })
+
+  const doOpenEditor = eventCallbacks.useEventCallback((projectId: Project['id']) => {
+    setPage(projectId)
+  })
+
+  const doCloseProject = eventCallbacks.useEventCallback((project: Project) => {
+    const { id } = project
+
+    closeProjectMutation.mutate(project)
+    dispatchAssetEvent({ type: AssetEventType.closeProject, id })
+    removeLaunchedProject(project.id)
+
+    setPage(TabType.drive)
+  })
+
+  const doRemoveSelf = eventCallbacks.useEventCallback((project: Project) => {
+    dispatchAssetListEvent({ type: AssetListEventType.removeSelf, id: project.id })
+    doCloseProject(project)
+  })
+
+  const onSignOut = eventCallbacks.useEventCallback(() => {
+    setPage(TabType.drive)
+
+    launchedProjects.forEach(project => {
```

Comment:
prefer a `for...of...` loop, even if just for consistency

Reply:
I'm not sure that just banning `foreach` in favor of `forof` is a good idea... If we don't have strong arguments against using a particular syntax, it's up to the author of the PR, right?

Reply:
On the second thought we can simplify this to just `doCloseAllProject`